File: src/lib/notify.test.ts
import { afterEach, describe, expect, it, vi } from 'vitest'
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import {
  NOTIFY_MESSAGE_MAX_LEN,
  NOTIFY_TRUNCATION_SUFFIX,
  notify_user,
  notifyRecordInserted,
  scheduleBestEffortNotify,
  shouldSuppressBotNotification,
  truncateNotifyMessage,
} from './notify'
import { clearAccessTokenCacheForTests } from './qqbot'

afterEach(() => {
  clearAccessTokenCacheForTests()
  vi.restoreAllMocks()
})

const truncateCases = JSON.parse(
  readFileSync(
    path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '../../testdata/notify-truncate-cases.json',
    ),
    'utf8',
  ),
) as {
  max_len: number
  suffix: string
  cases: Array<{
    name: string
    length: number
    cjk: boolean
    truncated: boolean
  }>
}

describe('truncateNotifyMessage', () => {
  it('constants stay byte-identical to Go notify', () => {
    expect(NOTIFY_MESSAGE_MAX_LEN).toBe(truncateCases.max_len)
    expect(NOTIFY_TRUNCATION_SUFFIX).toBe(truncateCases.suffix)
  })

  it('passes shared fixtures', () => {
    for (const tc of truncateCases.cases) {
      const unit = tc.cjk ? '字' : 'a'
      const input = unit.repeat(tc.length)
      const out = truncateNotifyMessage(input)
      if (!tc.truncated) {
        expect(out, tc.name).toBe(input)
        continue
      }
      expect(out.length, tc.name).toBe(NOTIFY_MESSAGE_MAX_LEN)
      expect(out.endsWith(NOTIFY_TRUNCATION_SUFFIX), tc.name).toBe(true)
      expect(out.startsWith(unit.repeat(10)), tc.name).toBe(true)
    }
  })
})

describe('shouldSuppressBotNotification', () => {
  it('skips only when SUPPRESS_BOT_NOTIFICATION trim equals 1', () => {
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: '1' }),
    ).toBe(true)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: ' 1 ' }),
    ).toBe(true)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: '0' }),
    ).toBe(false)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: 'true' }),
    ).toBe(false)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: 'yes' }),
    ).toBe(false)
  })

  it('falls back to process.env when injected flag is empty', () => {
    // tests/setup.ts 设 SUPPRESS_BOT_NOTIFICATION=1；空注入应回退到 process.env
    expect(shouldSuppressBotNotification({})).toBe(true)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: '' }),
    ).toBe(true)
    expect(
      shouldSuppressBotNotification({ SUPPRESS_BOT_NOTIFICATION: '0' }),
    ).toBe(false)
  })
})

const sampleNumber = {
  id: '01900000-0000-7000-8000-000000000001',
  happened_at: '2026-07-31T12:00:00.000Z',
  numeric_value: '72.5',
  raw_content: null,
  tags: ['weight', 'morning'],
  objective_context: 'Scale reading',
  ai_analysis: 'Feeling lighter',
}

describe('scheduleBestEffortNotify', () => {
  it('runs task outside Next request scope (route unit tests / fallback)', async () => {
    const task = vi.fn().mockResolvedValue(undefined)
    scheduleBestEffortNotify(task)
    await vi.waitFor(() => expect(task).toHaveBeenCalledTimes(1))
  })
})

describe('notify_user', () => {
  it('skips when SUPPRESS_BOT_NOTIFICATION=1 even when channels are configured', async () => {
    const fetchMock = vi.fn()
    await notify_user('hello', {
      env: {
        TELEGRAM_BOT_TOKEN: 't',
        TELEGRAM_USER_ID: '1',
        QQBOT_APP_ID: 'a',
        QQBOT_APP_SECRET: 's',
        QQBOT_USER_OPENID: 'o',
        SUPPRESS_BOT_NOTIFICATION: '1',
      },
      fetch: fetchMock,
    })
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it('warns when no channels configured', async () => {
    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})
    const fetchMock = vi.fn()
    await notify_user('hello', {
      env: { SUPPRESS_BOT_NOTIFICATION: '0' },
      fetch: fetchMock,
    })
    expect(fetchMock).not.toHaveBeenCalled()
    expect(warn).toHaveBeenCalledWith('Notify skipped: no channels configured')
  })

  it('sends Telegram only when only Telegram is configured', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => ({ ok: true }),
    })
    await notify_user('hi', {
      env: {
        TELEGRAM_BOT_TOKEN: 'tok',
        TELEGRAM_USER_ID: '9',
        SUPPRESS_BOT_NOTIFICATION: '0',
      },
      fetch: fetchMock,
    })
    expect(fetchMock).toHaveBeenCalledTimes(1)
    expect(fetchMock.mock.calls[0][0]).toContain('api.telegram.org')
  })

  it('truncates overlong text before sending to any channel', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => ({ ok: true }),
    })
    const long = 'a'.repeat(5000)
    await notify_user(long, {
      env: {
        TELEGRAM_BOT_TOKEN: 'tok',
        TELEGRAM_USER_ID: '9',
        SUPPRESS_BOT_NOTIFICATION: '0',
      },
      fetch: fetchMock,
    })
    expect(fetchMock).toHaveBeenCalledTimes(1)
    const body = JSON.parse(String(fetchMock.mock.calls[0][1].body)) as {
      text: string
    }
    expect(body.text.length).toBe(NOTIFY_MESSAGE_MAX_LEN)
    expect(body.text.endsWith(NOTIFY_TRUNCATION_SUFFIX)).toBe(true)
  })

  it('sends QQ only when only QQ is configured', async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ access_token: 'qt', expires_in: 7200 }),
      })
      .mockResolvedValueOnce({
        ok: true,
        status: 200,
        text: async () => '{}',
        json: async () => ({}),
      })
    await notify_user('hi', {
      env: {
        QQBOT_APP_ID: 'a',
        QQBOT_APP_SECRET: 's',
        QQBOT_USER_OPENID: 'o',
        SUPPRESS_BOT_NOTIFICATION: '0',
      },
      fetch: fetchMock,
    })
    expect(fetchMock.mock.calls.some((c) => String(c[0]).includes('bots.qq.com'))).toBe(
      true,
    )
    expect(
      fetchMock.mock.calls.some((c) => String(c[0]).includes('/v2/users/')),
    ).toBe(true)
  })

  it('sends Telegram and QQ in parallel when both configured', async () => {
    let inflight = 0
    let maxInflight = 0
    const fetchMock = vi.fn().mockImplementation(async (url: string) => {
      inflight += 1
      maxInflight = Math.max(maxInflight, inflight)
      await new Promise((r) => setTimeout(r, 30))
      inflight -= 1
      if (String(url).includes('bots.qq.com')) {
        return {
          ok: true,
          status: 200,
          json: async () => ({ access_token: 'qt', expires_in: 7200 }),
        }
      }
      if (String(url).includes('/v2/users/')) {
        return {
          ok: true,
          status: 200,
          text: async () => '{}',
          json: async () => ({}),
        }
      }
      return {
        ok: true,
        status: 200,
        json: async () => ({ ok: true }),
      }
    })

    await notify_user('hi', {
      env: {
        TELEGRAM_BOT_TOKEN: 'tok',
        TELEGRAM_USER_ID: '9',
        QQBOT_APP_ID: 'a',
        QQBOT_APP_SECRET: 's',
        QQBOT_USER_OPENID: 'o',
        SUPPRESS_BOT_NOTIFICATION: '0',
      },
      fetch: fetchMock,
    })

    expect(maxInflight).toBeGreaterThanOrEqual(2)
    expect(
      fetchMock.mock.calls.some((c) => String(c[0]).includes('api.telegram.org')),
    ).toBe(true)
    expect(
      fetchMock.mock.calls.some((c) => String(c[0]).includes('/v2/users/')),
    ).toBe(true)
  })

  it('returns after timeout without throwing when a channel hangs', async () => {
    const fetchMock = vi.fn().mockImplementation(
      () => new Promise(() => {}),
    )
    const started = Date.now()
    await expect(
      notify_user('hi', {
        env: {
          TELEGRAM_BOT_TOKEN: 'tok',
          TELEGRAM_USER_ID: '9',
          SUPPRESS_BOT_NOTIFICATION: '0',
        },
        fetch: fetchMock,
        timeoutMs: 50,
      }),
    ).resolves.toBeUndefined()
    expect(Date.now() - started).toBeLessThan(500)
  })

  it('logs channel failure without throwing', async () => {
    const error = vi.spyOn(console, 'error').mockImplementation(() => {})
    const fetchMock = vi.fn().mockResolvedValue({
      ok: false,
      status: 401,
      json: async () => ({ ok: false, description: 'Unauthorized' }),
    })
    await expect(
      notify_user('hi', {
        env: {
          TELEGRAM_BOT_TOKEN: 't',
          TELEGRAM_USER_ID: '1',
          SUPPRESS_BOT_NOTIFICATION: '0',
        },
        fetch: fetchMock,
      }),
    ).resolves.toBeUndefined()
    expect(error).toHaveBeenCalled()
  })
})

describe('notifyRecordInserted', () => {
  it('skips when unconfigured and does not call fetch', async () => {
    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {})
    const fetchMock = vi.fn()
    await notifyRecordInserted(sampleNumber, {
      env: { SUPPRESS_BOT_NOTIFICATION: '0' },
      fetch: fetchMock,
    })
    expect(fetchMock).not.toHaveBeenCalled()
    expect(warn).toHaveBeenCalled()
  })

  it('formats then notifies via notify_user', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => ({ ok: true }),
    })
    await notifyRecordInserted(sampleNumber, {
      env: {
        TELEGRAM_BOT_TOKEN: 't',
        TELEGRAM_USER_ID: '1',
        SUPPRESS_BOT_NOTIFICATION: '0',
      },
      fetch: fetchMock,
    })
    expect(fetchMock).toHaveBeenCalledWith(
      expect.any(String),
      expect.objectContaining({
        body: expect.stringContaining('New record'),
      }),
    )
  })
})

File: faas/internal/notify/notify_test.go
package notify

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"unicode/utf8"
	"sync/atomic"
	"testing"
	"time"

	"github.com/mdk/digitaltwin2026/faas/internal/qqbot"
	"github.com/mdk/digitaltwin2026/faas/internal/record"
	"github.com/mdk/digitaltwin2026/faas/internal/telegram"
)

func TestMain(m *testing.M) {
	_ = os.Setenv("SUPPRESS_BOT_NOTIFICATION", "1")
	os.Exit(m.Run())
}

type truncateCases struct {
	MaxLen int `json:"max_len"`
	Suffix string `json:"suffix"`
	Cases  []struct {
		Name      string `json:"name"`
		Length    int    `json:"length"`
		CJK       bool   `json:"cjk"`
		Truncated bool   `json:"truncated"`
	} `json:"cases"`
}

func loadTruncateCases(t *testing.T) truncateCases {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	root := filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
	b, err := os.ReadFile(filepath.Join(root, "testdata", "notify-truncate-cases.json"))
	if err != nil {
		t.Fatal(err)
	}
	var cases truncateCases
	if err := json.Unmarshal(b, &cases); err != nil {
		t.Fatal(err)
	}
	return cases
}

func TestTruncateNotifyMessage(t *testing.T) {
	cases := loadTruncateCases(t)
	if NotifyMessageMaxLen != cases.MaxLen || NotifyTruncationSuffix != cases.Suffix {
		t.Fatalf("constants mismatch fixture: len=%d suffix=%q", cases.MaxLen, cases.Suffix)
	}
	for _, tc := range cases.Cases {
		unit := "a"
		if tc.CJK {
			unit = "字"
		}
		input := strings.Repeat(unit, tc.Length)
		out := TruncateNotifyMessage(input)
		if !tc.Truncated {
			if out != input {
				t.Fatalf("%s: want unchanged", tc.Name)
			}
			continue
		}
		if len([]rune(out)) != NotifyMessageMaxLen {
			t.Fatalf("%s: len=%d want %d", tc.Name, len([]rune(out)), NotifyMessageMaxLen)
		}
		if !strings.HasSuffix(out, NotifyTruncationSuffix) {
			t.Fatalf("%s: missing suffix", tc.Name)
		}
		if !strings.HasPrefix(out, strings.Repeat(unit, 10)) {
			t.Fatalf("%s: missing head", tc.Name)
		}
	}
}

func TestShouldSuppressBotNotification(t *testing.T) {
	cases := []struct {
		name string
		val  string
		want bool
	}{
		{name: "one", val: "1", want: true},
		{name: "trimmed_one", val: " 1 ", want: true},
		{name: "zero", val: "0", want: false},
		{name: "true", val: "true", want: false},
		{name: "yes", val: "yes", want: false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ShouldSuppressBotNotification(func(k string) string {
				if k == "SUPPRESS_BOT_NOTIFICATION" {
					return tc.val
				}
				return ""
			})
			if got != tc.want {
				t.Fatalf("val=%q got=%v want=%v", tc.val, got, tc.want)
			}
		})
	}

	// 空注入回退 process.env（TestMain 设 SUPPRESS=1）
	if !ShouldSuppressBotNotification(func(string) string { return "" }) {
		t.Fatal("expected fallback to process.env SUPPRESS=1")
	}
}

func TestNotifyUserSkipsWhenSuppressed(t *testing.T) {
	called := false
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		called = true
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "1"
				case "TELEGRAM_BOT_TOKEN":
					return "tok"
				case "TELEGRAM_USER_ID":
					return "1"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	n.NotifyUser("hello")
	if called {
		t.Fatal("should not call channels when SUPPRESS_BOT_NOTIFICATION=1")
	}
}

func TestNotifyUserWarnsWhenNoChannels(t *testing.T) {
	n := &Notifier{
		Getenv: func(k string) string {
			if k == "SUPPRESS_BOT_NOTIFICATION" {
				return "0"
			}
			return ""
		},
		Telegram: &telegram.Sender{Getenv: func(string) string { return "" }},
		Qqbot:    &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	// 不应 panic；未配置时只打 warn 日志
	n.NotifyUser("hello")
}

func TestNotifyUserTelegramOnly(t *testing.T) {
	var calls int32
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt32(&calls, 1)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "0"
				case "TELEGRAM_BOT_TOKEN":
					return "tok"
				case "TELEGRAM_USER_ID":
					return "9"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	n.NotifyUser("hi")
	if atomic.LoadInt32(&calls) != 1 {
		t.Fatalf("calls=%d", calls)
	}
}

func TestNotifyUserQQOnly(t *testing.T) {
	var tokenOK, sendOK int32
	tokenSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt32(&tokenOK, 1)
		_, _ = w.Write([]byte(`{"access_token":"qt","expires_in":7200}`))
	}))
	defer tokenSrv.Close()
	sendSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt32(&sendOK, 1)
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{}`))
	}))
	defer sendSrv.Close()

	env := func(k string) string {
		switch k {
		case "SUPPRESS_BOT_NOTIFICATION":
			return "0"
		case "QQBOT_APP_ID":
			return "a"
		case "QQBOT_APP_SECRET":
			return "s"
		case "QQBOT_USER_OPENID":
			return "o"
		}
		return ""
	}
	n := &Notifier{
		Getenv:   env,
		Telegram: &telegram.Sender{Getenv: func(string) string { return "" }},
		Qqbot: &qqbot.Sender{
			HTTPClient: sendSrv.Client(),
			TokenURL:   tokenSrv.URL,
			APIBases:   []string{sendSrv.URL},
			Getenv:     env,
		},
	}
	n.NotifyUser("hi")
	if atomic.LoadInt32(&tokenOK) < 1 || atomic.LoadInt32(&sendOK) < 1 {
		t.Fatalf("token=%d send=%d", tokenOK, sendOK)
	}
}

func TestNotifyUserBothParallel(t *testing.T) {
	var inflight, maxInflight int32
	tgSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		cur := atomic.AddInt32(&inflight, 1)
		for {
			old := atomic.LoadInt32(&maxInflight)
			if cur <= old || atomic.CompareAndSwapInt32(&maxInflight, old, cur) {
				break
			}
		}
		time.Sleep(30 * time.Millisecond)
		atomic.AddInt32(&inflight, -1)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer tgSrv.Close()

	tokenSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		cur := atomic.AddInt32(&inflight, 1)
		for {
			old := atomic.LoadInt32(&maxInflight)
			if cur <= old || atomic.CompareAndSwapInt32(&maxInflight, old, cur) {
				break
			}
		}
		time.Sleep(30 * time.Millisecond)
		atomic.AddInt32(&inflight, -1)
		_, _ = w.Write([]byte(`{"access_token":"qt","expires_in":7200}`))
	}))
	defer tokenSrv.Close()

	qqSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{}`))
	}))
	defer qqSrv.Close()

	env := func(k string) string {
		switch k {
		case "SUPPRESS_BOT_NOTIFICATION":
			return "0"
		case "TELEGRAM_BOT_TOKEN":
			return "tok"
		case "TELEGRAM_USER_ID":
			return "9"
		case "QQBOT_APP_ID":
			return "a"
		case "QQBOT_APP_SECRET":
			return "s"
		case "QQBOT_USER_OPENID":
			return "o"
		}
		return ""
	}
	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: tgSrv.Client(),
			APIBase:    tgSrv.URL,
			Getenv:     env,
		},
		Qqbot: &qqbot.Sender{
			HTTPClient: qqSrv.Client(),
			TokenURL:   tokenSrv.URL,
			APIBases:   []string{qqSrv.URL},
			Getenv:     env,
		},
	}
	n.NotifyUser("hi")
	if atomic.LoadInt32(&maxInflight) < 2 {
		t.Fatalf("maxInflight=%d want >=2", maxInflight)
	}
}

func TestNotifyUserTimeout(t *testing.T) {
	// 永不返回的 RoundTrip，避免 httptest.Close 等 handler Sleep
	block := make(chan struct{})
	t.Cleanup(func() { close(block) })

	n := &Notifier{
		Timeout: 50 * time.Millisecond,
		Telegram: &telegram.Sender{
			HTTPClient: &http.Client{Transport: roundTripFunc(func(*http.Request) (*http.Response, error) {
				<-block
				return nil, fmt.Errorf("cancelled")
			})},
			APIBase: "http://example.invalid",
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "0"
				case "TELEGRAM_BOT_TOKEN":
					return "tok"
				case "TELEGRAM_USER_ID":
					return "9"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	started := time.Now()
	n.NotifyUser("hi")
	if time.Since(started) > 500*time.Millisecond {
		t.Fatal("should return after parallel timeout")
	}
}

type roundTripFunc func(*http.Request) (*http.Response, error)

func (f roundTripFunc) RoundTrip(r *http.Request) (*http.Response, error) { return f(r) }

func TestNotifyUserLogsFailureWithoutPanic(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"ok":false,"description":"Unauthorized"}`))
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "0"
				case "TELEGRAM_BOT_TOKEN":
					return "t"
				case "TELEGRAM_USER_ID":
					return "1"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	n.NotifyUser("hi")
}

func TestNotifyRecordInsertedFormats(t *testing.T) {
	var body string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		body = string(raw)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "0"
				case "TELEGRAM_BOT_TOKEN":
					return "t"
				case "TELEGRAM_USER_ID":
					return "1"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	num := "72.5"
	n.NotifyRecordInserted(record.Record{
		ID:               "id-1",
		HappenedAt:       "2026-07-31T12:00:00.000Z",
		NumericValue:      &num,
		Tags:             []string{"weight"},
		ObjectiveContext: "Scale",
	})
	if !strings.Contains(body, "New record") {
		t.Fatalf("body: %s", body)
	}
}

func TestNotifySkipsWhenUnconfigured(t *testing.T) {
	called := false
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		called = true
		_, _ = fmt.Fprintf(w, `{"ok":true}`)
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				if k == "SUPPRESS_BOT_NOTIFICATION" {
					return "0"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(k string) string {
			if k == "SUPPRESS_BOT_NOTIFICATION" {
				return "0"
			}
			return ""
		}},
	}
	n.NotifyRecordInserted(record.Record{ID: "x", HappenedAt: "t", Tags: []string{"a"}, ObjectiveContext: "o"})
	if called {
		t.Fatal("should not call when unconfigured")
	}
}

func TestNotifyNumberBatchInsertedFormats(t *testing.T) {
	var body string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		body = string(raw)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer srv.Close()

	n := &Notifier{
		Telegram: &telegram.Sender{
			HTTPClient: srv.Client(),
			APIBase:    srv.URL,
			Getenv: func(k string) string {
				switch k {
				case "SUPPRESS_BOT_NOTIFICATION":
					return "0"
				case "TELEGRAM_BOT_TOKEN":
					return "t"
				case "TELEGRAM_USER_ID":
					return "1"
				}
				return ""
			},
		},
		Qqbot: &qqbot.Sender{Getenv: func(string) string { return "" }},
	}
	v1, v2 := "72.5", "36.8"
	n.NotifyNumberBatchInserted([]record.Record{
		{ID: "id-1", HappenedAt: "2026-08-05T10:00:00+08:00", NumericValue: &v1, ObjectiveContext: "Scale"},
		{ID: "id-2", HappenedAt: "2026-08-05T10:00:00+08:00", NumericValue: &v2, ObjectiveContext: "axillary"},
	})
	if !strings.Contains(body, "New numbers batch") {
		t.Fatalf("body: %s", body)
	}
	if !strings.Contains(body, "inserted: 2") {
		t.Fatalf("body: %s", body)
	}
	// 逐条：value/memo/tags
	if !strings.Contains(body, "72.5/Scale/") {
		t.Fatalf("body: %s", body)
	}
	if !strings.Contains(body, "36.8/axillary/") {
		t.Fatalf("body: %s", body)
	}
}

func TestNotifyNumberBatchInsertedSkipsEmpty(t *testing.T) {
	// len==0 早返回，nil Notifier / 无渠道也不会 panic 或发送
	var n *Notifier
	n.NotifyNumberBatchInserted(nil)
}

func TestNotifyNumberBatchInsertedTruncatesOversized(t *testing.T) {
	rows := make([]record.Record, 100)
	for i := range rows {
		v := "36.8"
		rows[i] = record.Record{
			ID:               fmt.Sprintf("id-%d", i),
			HappenedAt:       "2026-08-05T10:00:00+08:00",
			NumericValue:      &v,
			Tags:             []string{"vitals"},
			ObjectiveContext: fmt.Sprintf("axillary temperature reading %d", i),
		}
	}
	raw := telegram.FormatNumberBatchMessage(rows)
	if utf8.RuneCountInString(raw) <= NotifyMessageMaxLen {
		t.Fatalf("raw should exceed limit, runes=%d", utf8.RuneCountInString(raw))
	}
	out := TruncateNotifyMessage(raw)
	if utf8.RuneCountInString(out) != NotifyMessageMaxLen {
		t.Fatalf("truncated runes=%d want %d", utf8.RuneCountInString(out), NotifyMessageMaxLen)
	}
	if !strings.HasSuffix(out, NotifyTruncationSuffix) {
		t.Fatalf("missing suffix: %q", out[len(out)-20:])
	}
}
